[TDA-7] Add functionality: "Tasks"

timesplash opened the pull request:
### TDA-7 Task Add functionality: "Tasks"

[Task link](https://todoapp.youtrack.cloud/issue/TDA-7/Add-functionality-Tasks)  
[Tester telegram link](https://t.me/aeDeaf)

### Short task description
Add task addition and fetching functionality to the app

### Readiness conditions
Passed unit-tests, manual testing

- [x] Unit-tests

timesplash committed:
[TDA-6] Translate app to english (#5)

* [TDA-6] Translate app to english
commons translation

* [TDA-6] Translate app to english
emulator translations

* [TDA-6] Translate app to english
ToDoApp-main translations, additions of forgotten commentaries
translation of description and pull request template

(cherry picked from commit 07f7ae8dc47867f2643030a9e7d10bfc74ec3f78)

timesplash committed:
[TDA-7] Add functionality: "Tasks"
deleted unnecessary import

timesplash committed:
[TDA-7] Add functionality: "Tasks"
added new request DTO + Kafka topic and Request Type to common

timesplash committed:
[TDA-7] Add functionality: "Tasks"
small changes to AddTaskRequestDTO + required stuff to todoapp-main

timesplash committed:
Merge branch 'master' into TDA-7

# Conflicts:
#	pull_request_template.md
#	todoapp-main/src/main/java/ru/todoapp/configuration/KafkaConfig.java

timesplash committed:
[TDA-7] Add functionality: "Tasks"
changed String to ZonedDateTime in AddTaskRequestDTO and subsequent changes to TaskService

timesplash committed:
[TDA-7] Add functionality: "Tasks"
fix to one untranslated parameter, things to add new tasks in emulator

timesplash committed:
[TDA-7] Add functionality: "Tasks"
Im a dumbass who forgot to add topic and factories to KafkaConfig in emulator

timesplash committed:
[TDA-7] Add functionality: "Tasks"
Im a dumbass who wrote a wong comment in javadoc

timesplash committed:
[TDA-7] Add functionality: "Tasks"
So, "fixed" unDeserializable ZonedDateTime by changing it to String in AddTaskRequestDTO. Changed the order of saving to requestRepository in Task and User Services so now only if no unprocessed exceptions happens it will save to RequestRepository.

timesplash committed:
[TDA-7] Add functionality: "Tasks"
intermediate version

timesplash committed:
[TDA-7] Add functionality: "Tasks"
intermediate version

timesplash committed:
[TDA-7] Add functionality: "Tasks"
intermediate version

timesplash committed:
[TDA-7] Add functionality: "Tasks"
DTOs for task fetching + KafkaConfigs for that

timesplash committed:
[TDA-7] Add functionality: "Tasks"
Prepped Listeners for the future
Entities and repository in first iteration might will change in the future

timesplash committed:
[TDA-7] Add functionality: "Tasks"
Yaponu mat' ego rot blyat'

timesplash committed:
[TDA-7] Add functionality: "Tasks"
commit for PR

timesplash committed:
[TDA-7] Add functionality: "Tasks"
return of the lost PULL_REQUEST_TEMPLATE.md

aeDeaf commented on todoapp-common/src/main/java/ru/todoapp/model/dto/AddTaskRequestDTO.java:
```diff
@@ -0,0 +1,28 @@
+    /**
+     * datetime, when the task is planned to take place, in ISO 8601 format (yyyy-MM-ddThh:mm:ss+hh)
+     */
+    private String datetime;
```
Why you use here a string instead of long?
I thought we both agree, that we will use the UNIX time without a timezone

timesplash replied:
We have tried to parse it to Instant (the field looked like: Instant datetime) and we were getting long from emulator, but I thought it was not super comfortable and changed it back in EmolatorDTO to String. 

Then I realized that we will not be able to have similar intafaces between AddTaskDTO (Instant datetime) and FetchTasksRequestDTO (String beginDate/endDate) and decided to go back to String.

aeDeaf commented on todoapp-common/src/main/java/ru/todoapp/model/dto/FetchTasksRequestDTO.java:
```diff
@@ -0,0 +1,27 @@
+    /**
+     * Date from which we need the list of tasks
+     */
+    private String beginDate;
```
I think we should use here long values too

aeDeaf commented on todoapp-main/src/main/java/ru/todoapp/configuration/KafkaConfig.java:
```diff
@@ -10,9 +10,7 @@
-import ru.todoapp.model.dto.PingRequestDTO;
-import ru.todoapp.model.dto.RequestResultDTO;
-import ru.todoapp.model.dto.RegisterRequestDTO;
+import ru.todoapp.model.dto.*;
```
Don't use wildcard imports (with '*'). You cans use them only on tests.
Have you applied a codestyle file from this repo to your's idea?

timesplash replied:
idea_set_code_style.xml?
I think we did... then we were merging this changes to that branch

aeDeaf commented on todoapp-main/src/main/java/ru/todoapp/configuration/KafkaConfig.java:
```diff
@@ -67,13 +65,43 @@ public NewTopic requestResultTopic() {
         return new NewTopic(KafkaTopics.REGISTRATION_TOPIC, 1, (short) 1);
     }
 
+    /**
+     * Kafka topic creation for adding tasks
+     *
+     * @see ru.todoapp.model.dto.AddTaskRequestDTO
+     */
+    @Bean
+    public NewTopic addTaskTopic() {
+        return new NewTopic(KafkaTopics.ADD_TASK_TOPIC, 1, (short) 1);
+    }
+
+    /**
+     * Kafka topic creation for fetching tasks request
+     *
+     * @see ru.todoapp.model.dto.FetchTasksRequestDTO
+     */
+    @Bean
+    public NewTopic fetchTasksRequestTopic() {
+        return new NewTopic(KafkaTopics.FETCH_REQUEST_TASKS_TOPIC, 1, (short) 1);
+    }
+
+    /**
+     * Kafka topic creation for fetching tasks response
+     *
+     * @see ru.todoapp.model.dto.FetchTasksResponseDTO
+     */
+    @Bean
+    public NewTopic fetchTasksResponseTopic() {
+        return new NewTopic(KafkaTopics.FETCH_RESPONSE_TASKS_TOPIC, 1, (short) 1);
+    }
+
```
I think we should use a constants instead of 1 in this whole class. Probably I just forgot to replace here this numbers with constants.

timesplash replied:
Ok will change

aeDeaf commented on todoapp-main/src/main/java/ru/todoapp/model/SaveTaskEntity.java:
```diff
@@ -0,0 +1,23 @@
+ * @param userUUID    - userUUID of user that planned task
+ * @param datetime    - timestamp then the task planned to take place
+ */
+public record SaveTaskEntity(String description, String userUUID, OffsetDateTime datetime) {
```
You don't need two classes to represent one entity. Instead of this, you can create one class with the same fields as in the corresponding table. After that, you just fill the only needed fields. For example, when you create entity before saving it, you just don't fill the ID field ('cause you don't know it value yet)

timesplash replied:
Ok

aeDeaf commented on todoapp-main/src/main/java/ru/todoapp/repository/TaskRepository.java:
```diff
@@ -0,0 +1,42 @@
+        jdbcClient.sql(SAVE_NEW_TASK).params(params).update();
+    }
+
+    public List<TaskEntity> getTasksList(GetTaskRequestEntity getTaskRequestEntity) {
```
you don't need to use a class parameter in this method. You can just pass user UUID, begin and end date as a parameters, so the class GetTaskRequestEntity become redundant

timesplash replied:
Ok

Addm1X commented on todoapp-main/build.gradle:
```diff
@@ -13,6 +13,7 @@ repositories {
 
 dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-jdbc'
+    implementation 'com.fasterxml.jackson.datatype:jackson-datatype-jsr310:2.16.1'
```
Whats that for? I removed from dependency and nothing changed

aeDeaf replied:
This dependency improve datetime deserialization with jackson and takes an effect only in runtime.

Addm1X commented on todoapp-main/src/test/java/ru/todoapp/service/TaskServiceTest.java:
```diff
@@ -0,0 +1,264 @@
+     */
+    @Test
+    void handleAdditionSuccessTest() {
+        when(userRepositoryMock.exists(any(String.class))).thenReturn(true);
```
Good practise to comment given/when/then between 3 logical steps of the test.

timesplash replied:
Ok

Addm1X commented on todoapp-main/src/main/java/ru/todoapp/service/UserService.java:
```diff
@@ -44,6 +42,8 @@ public void handle(RegisterRequestDTO registerRequestDTO) {
             sendRequestResultDTO(registerRequestDTO.getRequestUUID(), RequestStatus.SUCCESS, null);
         }
+
+        requestRepository.save(RequestEntity.of(registerRequestDTO));
```
Why you put this at the end of the method? I understand that it's changed nothing, but is not it more logical to save request right it has been received?

timesplash replied:
Yes and no, If something fails completely Kafka with its current config will continue sending me request that will fail, the main body of the request will not be saved to its own table (user/task) but it will fuck up our request table untill it will timeout + @aeDeaf said that that table and everything with it will go at the end anyway

Addm1X commented on todoapp-main/src/main/java/ru/todoapp/repository/TaskRepository.java:
```diff
@@ -0,0 +1,42 @@
+        jdbcClient.sql(SAVE_NEW_TASK).params(params).update();
+    }
+
+    public List<TaskEntity> getTasksList(GetTaskRequestEntity getTaskRequestEntity) {
```
We agreed on our conventions, that all public methods should have a javadoc

timesplash replied:
Yep, sorry forgot a couple

Addm1X commented on todoapp-main/src/main/java/ru/todoapp/model/TaskEntity.java:
```diff
@@ -0,0 +1,11 @@
+
+import java.time.OffsetDateTime;
+
+public record TaskEntity(long id, String description, OffsetDateTime datetime) {
```
Total lack of javadocs

Addm1X commented on todoapp-emulator/src/main/java/emulator/ru/todoapp/model/FetchTasksRequestEmulatorDTO.java:
```diff
@@ -0,0 +1,4 @@
+package emulator.ru.todoapp.model;
+
+public record FetchTasksRequestEmulatorDTO(String userUUID, String beginDate, String endDate) {
```
Lack of javadocs

Addm1X commented on todoapp-common/src/main/java/ru/todoapp/model/dto/TaskDTO.java:
```diff
@@ -0,0 +1,7 @@
+import lombok.Builder;
+
+@Builder
+public record TaskDTO(long id, String description, String datetime) {
```
Lack of javadocs

Addm1X commented on todoapp-main/src/main/java/ru/todoapp/service/TaskService.java:
```diff
@@ -0,0 +1,175 @@
+     *
+     * @param addTaskRequestDTO - the object checked for unfilled fields
+     */
+    private List<String> allUnfilledFields(AddTaskRequestDTO addTaskRequestDTO) {
```
1. better name it like `getAllUnfilledFields`, it's a method, not a field
2. as i can see you use those strings for messaging, but it's not necessarily. You can override `toString()` method in our class and bypass object to the message. Overriden method does all job for you.
3. To check whatever it's all field are fulfilled i suggest to create some kind of DTO validator, or better yet create abstract boolean method in `RequestDTO`, implemented in each DTO that returns all fields are valid or not. That way, we don't new boilerplate overriden methods in the future.

aeDeaf replied:
I think we shouldn't spend time on DTO validators right now, 'cause we can use hibernate validators instead. This validators (without hibernate) are only for testing purposes.

Addm1X replied:
Cute enthusiasm, but ok. That said, only points 1 and 2 remains from my comment.

timesplash replied:
I did not quite get what do you mean by overriding `toString()` to return an empty field names. If I understood you correctly I will create a new method in the classes that will be `getAllUnfilledFields()` that will return a list of Strings containing names of the fields, I dont think I need to override the `toString()` for that... If that is not what you meant I quite dont understand how `toString()`, that will return `String` containing lets say every field value of an object, will help me in determining the missing ones, I believe it will be quite a pain in the ass to work with that String value

Addm1X replied:
`toString()` has nothing to do with determining empty fields. This thing helps you to avoid verbose passing name fields, thats it. Override object `toString()`, then call it in `sendRequestResultDTO` for example, in message argument. Profit.

timesplash replied:
I think I get it, but on second reread of your first comment I think we are talking about different things

You suggest `toString()` method to avoid:
`var result = SomethingDTO.builder()...
SomethingDTOKafkaTemplate.send(..., result);`
Correct?

I was thinking of that suggestion on the `allUnfilledFields()` method, that method only for creation failmessage of failed request in case fail accured then one or more fields are empty

Addm1X commented on todoapp-main/src/main/java/ru/todoapp/service/TaskService.java:
```diff
@@ -0,0 +1,175 @@
+     * method should be used in case then you already know that at least one of them is incorrect
+     */
+    private String incorrectDateOrDates(FetchTasksRequestDTO fetchTasksRequestDTO) {
+        if (!DateTimeUtils.isCorrectDateFormat(fetchTasksRequestDTO.getBeginDate())
```
same here for DTO validator

timesplash committed:
[TDA-7] Add functionality: "Tasks"
minor fixes

timesplash committed:
[TDA-7] Add functionality: "Tasks"
major fixes and changes:
a bunch of forgotten comments, changes to how date/datetime variables are transported and subsequent changes to use-cases

timesplash committed:
[TDA-7] Add functionality: "Tasks"
last forgotten comments (I hope so)

timesplash committed:
[TDA-7] Add functionality: "Tasks"
minor changes

timesplash committed:
[TDA-7] Add functionality: "Tasks"
minor changes
